### ISSUE TITLE
The learning rule

### ISSUE THREAD

**Issue by uestcwangxiao** [EPFL-LCN/pub-illing2021-neurips#1]
Hi dear author,
 Are you using the local hebbian-like learning rule in your model training. Because in your main trainning  code, instead of your CLAPP rules   (main_vision.py ) your are still using loss_for_grads.backward(retain_graph=True) to train your model. Maybe there was something I was missing. In your model_utils.py, you did the comparision  between CLAPP and BP, but do you use CLAPP to do the actual training?

**Comment by berndilling** [comment]
Hi, thanks for your interest.

**TL;DR:** our `.backward()` implementation is equivalent with the local CLAPP rules, see README

For all CLAPP simulations, we use the local CLAPP rules for training, and no backpropagation between layers. However, we indeed _implement the local CLAPP rules using PyTorch's_ `.backward()` function, i.e. as the (auto-differentiated) gradients of the CLAPP loss Eq. 3 at every layer. Now for _CLAPP_ (but not for BP or Hinge Loss CPC), _backpropagation between layers is avoided through gradient blocking_, i.e. PyTorch's `.detach()` function. We actually double checked the equivalence of updates of the autodiff implementation (using `.backward()`) and a direct implementation of the local CLAPP rules Eqs. 6-8, as described in the README.

I hope this answers your question.

**Comment by uestcwangxiao** [comment]
Thanks for your quick reply, it's really helpful. The README is very clearly to decribe the process.
